### DISCUSSION TITLE
Unsupported Server Component type: undefined

### DISCUSSION THREAD

**Discussion by romanusname** [vercel/next.js#45347] [Help]
Hey! I ran into some weird behavior in Next js 13 while trying to do a simple action.
I have prepared the most simple example showing the problem. I hope you can help me figure out why this is happening.

There are 3 components.

**page:**
```
import {Container} from '../components/Container'

export default function index() {
  return (
    <Container>
        <Container.Window>
            <h1>12345</h1>
        </Container.Window>
    </Container>
  )
}
```
**Container:**
```
'use client';
import { Window } from './Window';

export const Container = ({ children }) => {
  return children;
};

Container.Window = Window;
```

**Window:**
```
"use client";
export const Window = ({children})=>{
  return children
}
```

**Page server component. The Container and Window components are client-side. Container imports and exports Window.**
I am getting "Unsupported Server Component type: undefined" error

Mistake can be avoided

**Import the Window component separately.**
Works, no errors

**Make the Container component a server component.**
Works, no errors

**Make the page component client-side.**
Works, no errors

But I think the first option is the most convenient. Would love to make it work
https://stackblitz.com/edit/nextjs-fwuztm?file=app%2Fpage.js,components%2FContainer.js,components%2FWindow.js - example

**Comment by IlyaSmagin** [comment]
As far as I understand, you can't import and use client-side compound components with dot notation from server-side components.
You can look for reference issue here https://github.com/vercel/next.js/issues/41940
Explanation as of why this is "intended" behavior (from Next.js contributor): https://github.com/vercel/next.js/issues/44030#issuecomment-1542597082

**Comment by RomainGueffier** [comment]
As I understand by running into this problem in my monorepo next js component library, import a compound component using 'use client' in root component file or barrel files does not work in next js. 
![Screenshot 2024-06-06 180818](https://github.com/vercel/next.js/assets/71011271/689f8f7a-4c98-4a94-ae72-90da6e150346)


To fix it, I need to add 'use client' in the page:
```tsx
'use client" // necessary or error is thrown
import { AgencyPage003 } from '@/lib/agencies'

export default function Page(){
  return (
   <AgencyPage003>
    <AgencyPage003.Header />
    <AgencyPage003.Content />
   </AgencyPage003>
  )
}
```

Will it work with this syntax ? Following related issues and discussions.. seems like yes
```tsx
// 'use client' not needed
import { AgencyPage003, AgencyPage003Header, AgencyPage003Body  } from '@lib/agencies'

export default function Page(){
  return (
   <AgencyPage003>
    <AgencyPage003Header />
    <AgencyPage003Body />
   </AgencyPage003>
  )
}
```

Before I dive into a full refactoring 😢, which syntax is considered best practice (bit out of topic) ? Shadcn ui, headless ui use the later for example. I guess it's also easier for bundlers to reduce unused modules?

Thanks to any feedback ✌🏼
